python/src/fast.cpp

```diff
@@ -56,4 +56,45 @@ void init_extensions(py::module_& parent_module) {
         Returns:
             array: The output array.
       )pbdoc");
+
+  m.def(
+      "scaled_dot_product_attention",
+      [](const array& q,
+         const array& k,
+         const array& v,
+         const float scale,
+         const std::optional<array>& mask,
+         const StreamOrDevice& s) {
+        return fast::scaled_dot_product_attention(q, k, v, scale, mask, s);
+      },
+      "q"_a,
+      "k"_a,
+      "v"_a,
+      py::kw_only(),
+      "scale"_a,
+      "mask"_a = none,
+      "stream"_a = none,
+      R"pbdoc(
+                  scaled_dot_product_attention(q: array, k: array, v: array, *, scale: float, /,  mask: Union[None, array] = None, stream: Union[None, Stream, Device] = None) -> array
+
+            A parallelized implementation of multi-head attention: O = softmax(Q @ K.T, dim=-1) @ V.
+            Supports Multi-Head Attention (see https://arxiv.org/abs/1706.03762),
+            Grouped Query Attention (https://arxiv.org/abs/2305.13245),
+            and Multi-Query Attention (https://arxiv.org/pdf/1911.02150.pdf).
```

Comment:
```suggestion
            and Multi-Query Attention (https://arxiv.org/abs/1911.02150).
```

Reply:
Would you also mind adding something about the fact that the softmax is done in higher precision regardless of the precision of the inputs?

E.g.:

The ``softmax`` is always performed in ``float32`` and then cast back to the input dtype before multiplying with ``V``.

Reply:
Yes, will do

Reply:
looks like Atila covered this, marking as resolved.  thanks!